File: src/main/java/fyordo/lifeagragator/med/tag/TagService.java
package fyordo.lifeagragator.med.tag;

import com.querydsl.core.types.dsl.BooleanExpression;
import fyordo.lifeagragator.med.base.exceptions.ModelNotFoundException;
import fyordo.lifeagragator.med.base.utils.WorkspaceUtils;
import fyordo.lifeagragator.med.tag.request.TagCreateRequest;
import fyordo.lifeagragator.med.tag.request.TagFilter;
import fyordo.lifeagragator.med.tag.request.TagUpdateRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;

@Service
@RequiredArgsConstructor
public class TagService {
    private final TagRepository tagRepository;

    public List<Tag> getTags(TagFilter tagFilter){
        return (List<Tag>)tagRepository
                .findAll(generateFilter(tagFilter));
    }

    private BooleanExpression generateFilter(TagFilter tagFilter) {
        QTag qTag = QTag.tag;
        BooleanExpression result = qTag.isNotNull();

        if (tagFilter.getOnlyMy()){
            result = result.and(qTag.createdUserId.eq(WorkspaceUtils.getUserId()));
        }

        if (tagFilter.getSearch() != null){
            result = result.and(qTag.title.containsIgnoreCase(tagFilter.getSearch()));
        }

        return result;
    }


    public Tag getTagById(Long id){
        return tagRepository
                .findById(id)
                .filter((Tag tag) -> Objects.equals(tag.getCreatedUserId(), WorkspaceUtils.getUserId()))
                .orElseThrow(ModelNotFoundException::new);
    }

    public Tag createTag(TagCreateRequest data){
        Tag tag = new Tag(data);
        tag.setCreatedUserId(WorkspaceUtils.getUserId());
        return tagRepository.save(tag);
    }

    public Tag updateTag(TagUpdateRequest data){
        getTagById(data.getId());

        Tag tag = new Tag(data);
        tag.setCreatedUserId(WorkspaceUtils.getUserId());
        return tagRepository.save(tag);
    }

    public void deleteTagById(Long id){
        Tag tag = getTagById(id);
        tagRepository.delete(tag);
    }
}

File: src/main/kotlin/fyordo/lifeagragator/med/tag/TagService.kt
package fyordo.lifeagragator.med.tag

import fyordo.lifeagragator.med.base.ModelNotFoundException
import fyordo.lifeagragator.med.tag.dto.TagCreateRequest
import fyordo.lifeagragator.med.tag.dto.TagUpdateRequest
import kotlinx.coroutines.*
import mu.KotlinLogging
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import kotlin.math.log

@Service
class TagService(
    private val tagRepository: TagRepository,
) {
    private val logger = KotlinLogging.logger {}

    fun findAll(): List<Tag> =
        tagRepository.findAll()

    fun findById(id: Long): Tag =
        tagRepository.findById(id)
            .orElseThrow { ModelNotFoundException() }

    fun createTag(data: TagCreateRequest): Tag =
        tagRepository.save(Tag(data))

    fun updateTag(data: TagUpdateRequest) =
        tagRepository.save(Tag(data))

    fun deleteTag(id: Long): Unit =
        tagRepository.delete(findById(id))

    suspend fun deleteTags(ids: List<Long>): Unit = coroutineScope{
        launch (CoroutineName("MULTIPLE_TAG_DELETE") + Dispatchers.IO + SupervisorJob()) {
            for (id in ids){
                launch {
                    if (id == 291L) {
                        throw ModelNotFoundException()
                    }
                    deleteTag(id)
                    logger.info("DELETED_TAG_$id")
                }
            }
        }
    }
}